File: Backtrack/*Remove Invalid Parentheses/solution_dfs.ts
function removeInvalidParentheses(s: string): string[] {
    let stack = []
    let notValidOpen = 0
    let notValidClose = 0
    for(let l of s) {
        if(l === "(") {
            stack.push(l)
        } else if(l === ")") {
            if(!stack.length) notValidClose++
            else stack.pop()
        }
    }
    notValidOpen = stack.length
    let result = new Set()
    let dfs = (idx, discardOpen, discardClose, stack, currentString) => {
        if(discardOpen < 0 || discardClose < 0) return
        if(idx >= s.length) {
            if(discardOpen === 0 && discardClose === 0 && stack === 0) result.add(currentString)
            return
        }
        
        if(s[idx] === "(") {
            dfs(idx + 1, discardOpen, discardClose, stack + 1, currentString + "(")
            dfs(idx + 1, discardOpen - 1, discardClose, stack, currentString)
        } else if(s[idx] === ")") {
            dfs(idx + 1, discardOpen, discardClose - 1, stack, currentString)
            if(stack > 0) {
                dfs(idx + 1, discardOpen, discardClose, stack - 1, currentString + ")")
            } 
        } else {
            dfs(idx + 1, discardOpen, discardClose, stack, currentString + s[idx])
        }
    }
    
    dfs(0, notValidOpen, notValidClose, 0, "")
    return Array.from(result) as string[]
};

File: Backtrack/*Remove Invalid Parentheses/solution_dfs.py
class Solution:
    def removeInvalidParentheses(self, s: str) -> List[str]:
        invalidOpen = 0
        invalidClose = 0
        stack = 0
        for l in s:
            if l == ")":
                if stack == 0: invalidClose += 1
                else: stack -= 1
            if l == "(":
                stack += 1
        invalidOpen = stack
        
        result = set()
        
        def dfs(idx, invalidOpen, invalidClose, stack, current):
            if invalidOpen < 0 or invalidClose < 0 or stack < 0:
                return
            if idx >= len(s):
                if invalidOpen == 0 and invalidClose == 0 and stack == 0:
                    result.add(current)
                    return
                else:
                    return

            if s[idx] == ")":
                dfs(idx + 1, invalidOpen, invalidClose - 1, stack, current)
                dfs(idx + 1, invalidOpen, invalidClose, stack - 1, current + s[idx])
            elif(s[idx] == "("):
                dfs(idx + 1, invalidOpen - 1, invalidClose, stack, current)
                dfs(idx + 1, invalidOpen, invalidClose, stack + 1, current + s[idx])
            else:
                dfs(idx + 1, invalidOpen, invalidClose, stack, current + s[idx])
                
        
        dfs(0, invalidOpen, invalidClose, 0, "")
        return list(result)
                
                    
                    